kbmd/piv.c

```diff
@@ -0,0 +1,1229 @@
+	}
+
+	if (BN_to_ASN1_INTEGER(serial, serial_asn1) == NULL) {
+		make_sslerrf(ret, "BN_to_ASN1_INTEGER",
+		    "setting certificate serial");
+		goto done;
+	}
+
+	if (X509_set_serialNumber(cert, serial_asn1) != 1) {
+		make_sslerrf(ret, "X509_set_serialNumber",
+		    "setting certificate serial");
+	}
+
+done:
+	BN_free(serial);
+	ASN1_INTEGER_free(serial_asn1);
+	return (ret);
+}
+
+static errf_t *
+set_subj_and_issuer(struct piv_token *restrict pk, X509 *restrict cert,
+    const char *name)
+{
+	errf_t *ret = ERRF_OK;
+	X509_NAME *subj = NULL;
+	const char *guidhex = piv_token_guid_hex(pk);
+
+	if ((subj = X509_NAME_new()) == NULL) {
+		make_sslerrf(ret, "X509_NAME_new",
+		    "setting subject and issuer");
+		goto done;
+	}
+
+	if (X509_NAME_add_entry_by_NID(subj, NID_title, MBSTRING_ASC,
+	    (unsigned char *)name, -1, -1, 0) != 1) {
+		make_sslerrf(ret, "X509_NAME_add_entry_by_NID",
+		    "setting subject and issuer");
+		goto done;
+	}
+
+	if (X509_NAME_add_entry_by_NID(subj, NID_commonName, MBSTRING_ASC,
+	    (unsigned char *)guidhex, -1, -1, 0) != 1) {
+		make_sslerrf(ret, "X509_NAME_add_entry_by_NID",
+		    "setting subject and issuer");
+		goto done;
+	}
+
+	/* XXX: Set OU or organization name? */
+
+	if (X509_set_subject_name(cert, subj) != 1) {
+		make_sslerrf(ret, "X509_set_subject_name",
+		    "setting subject and issuer");
+		goto done;
+	}
+
+	if (X509_set_issuer_name(cert, subj) != 1) {
+		make_sslerrf(ret, "X509_set_issuer_name",
+		    "setting subject and issuer");
+	}
+
+done:
+	X509_NAME_free(subj);
+	return (ret);
+}
+
+static errf_t *
+set_extensions(X509 *cert, const char *basic, const char *ku)
+{
+	errf_t *ret = ERRF_OK;
+	X509V3_CTX x509ctx;
+	X509_EXTENSION *ext = NULL;
+
+	X509V3_set_ctx_nodb(&x509ctx);
+	X509V3_set_ctx(&x509ctx, cert, cert, NULL, NULL, 0);
+
+	if ((ext = X509V3_EXT_conf_nid(NULL, &x509ctx, NID_basic_constraints,
+	    (char *)basic)) == NULL) {
+		make_sslerrf(ret, "X509V3_EXT_conf_nid", "setting extensions");
+		goto done;
+	}
+	X509_add_ext(cert, ext, -1);
+	X509_EXTENSION_free(ext);
+
+	if ((ext = X509V3_EXT_conf_nid(NULL, &x509ctx, NID_key_usage,
+	    (char *)ku)) == NULL) {
+		make_sslerrf(ret, "X509V3_EXT_conf_nid", "setting extensions");
+	}
+	X509_add_ext(cert, ext, -1);
+
+done:
+	X509_EXTENSION_free(ext);
+	return (ret);
+}
+
+static errf_t *
+set_pubkey_rsa(EVP_PKEY *restrict pkey, struct sshkey *restrict pub,
+    enum sshdigest_types *restrict algp, int *restrict nidp)
+{
+	errf_t *ret = ERRF_OK;
+	RSA *copy = NULL;
+
+	if ((copy = RSA_new()) == NULL) {
+		make_sslerrf(ret, "RSA_new", "setting public key");
+		goto done;
+	}
+
+	if ((copy->e = BN_dup(pub->rsa->e)) == NULL ||
+	    (copy->n = BN_dup(pub->rsa->n)) == NULL) {
+		make_sslerrf(ret, "BN_dup", "setting public key");
+		goto done;
+	}
+
+	if (EVP_PKEY_assign_RSA(pkey, copy) != 1) {
+		make_sslerrf(ret, "EVP_PKEY_assign_RSA", "setting public key");
+		goto done;
+	}
+
+	*nidp = NID_sha256WithRSAEncryption;
```

Comment:
It should be fine -- this is for creating the certificates on the PIV token. Once they're created, the PIV token is effectively sealed -- that is we cannot add or alter the certificates. The only thing that can be done is to factory reset (and erase everything) and re-initialize the PIV token. Presumably if the args were an issue, they'd have to go through that step regardless (and see below) might even have to just replace the PIV token outright.

Reply:
Okay. I always worry about algorithm agility, but if this involves PIV standards, then it's their problem per se, not ours.

Reply:
There are others options available, but of what is defined, these are the best ones currently available, and we didn't want to go through the effort of wiring up all the pieces through Triton to have an operator pick a poor algorithm. Since it's also something that can't be changed once the token is initialized, if newer/better ones do come out at some point in the future, we'd need to push a new update (and it's only something that happens during CN setup, so a new PI shouldn't be a huge issue).